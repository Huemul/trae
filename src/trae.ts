import merge from 'lodash/merge';

import createRequestBody from '../lib/create-request-body'
import createResponse from '../lib/create-response';
import { format as formatUrl } from '../lib/url';
import { TraeSettings, InstanceConfig } from './types';

const defaults: RequestInit = {
  headers: { 'Content-Type': 'application/json' },
};

function createTrae(providedConf?: Partial<TraeSettings>) {
  const config: TraeSettings = Object.freeze(merge(defaults, {
    before: (conf: RequestInit) => conf,
    onResolve: (item: unknown) => Promise.resolve(item),
    onReject: (err: unknown) => Promise.reject(err),
    ...providedConf,
  }));

  function request(endpoint: string, settings: TraeSettings) {
    // TODO: We should extract some attributes to avoid exposing unnecessary
    //       data to the before middleware and fetch function.
    //       Example: 'middleware' attribute.

    const url = formatUrl(settings.url, endpoint, settings.params);

    return fetch(url, config.before(settings))
      .then((res: any) => createResponse(res, settings))
      .then(config.onResolve)
      .catch(config.onReject);
  }

  function create(instanceConfig: InstanceConfig) {
    return createTrae(merge(config, instanceConfig));
  }

  function get(endpoint: string, requestConfig: RequestInit = {}) {
    const settings: TraeSettings = merge({}, config, requestConfig); 
    return request(endpoint, { ...settings, method: 'GET' });
  }

  function remove(endpoint: string, requestConfig: RequestInit = {}) {
    const settings: TraeSettings = merge({}, config, requestConfig); 
    return request(endpoint, { ...settings, method: 'DELETE' });
  }

  function head(endpoint: string, requestConfig: RequestInit = {}) {
    const settings: TraeSettings = merge({}, config, requestConfig); 
    return request(endpoint, { ...settings, method: 'HEAD' });
  }

  function post(
    endpoint: string,
    data: any = {},
    requestConfig: RequestInit = {},
  ) {
    const settings: TraeSettings = merge({}, config, requestConfig); 
    const body = createRequestBody(data, settings);
    return request(endpoint, { ...settings, method: 'POST', body });
  }

  function put(
    endpoint: string,
    data: any = {},
    requestConfig: RequestInit = {},
  ) {
    const settings: TraeSettings = merge({}, config, requestConfig);
    const body = createRequestBody(data, settings);
    return request(endpoint, { ...settings, method: 'PUT', body });
  }

  function patch(
    endpoint: string,
    data: any = {},
    requestConfig: RequestInit = {},
  ) {
    const settings: TraeSettings = merge({}, config, requestConfig);
    const body = createRequestBody(data, settings);
    return request(endpoint, { ...settings, method: 'PATCH', body });
  }

  return {
    create,
    get,
    delete: remove,
    head,
    post,
    put,
    patch,
    config,
  };
}

export default createTrae();
