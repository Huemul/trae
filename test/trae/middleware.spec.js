/* global describe it expect afterEach */

import fetchMock from 'fetch-mock';
import trae      from '../../lib';

afterEach(() => {
  fetchMock.restore();
});

const TEST_URL = 'http://localhost:8080/api';

describe('trae - middleware', () => {

  it('sets the middlewares', () => {
    const apiFoo    = trae.create();
    const identity  = response => response;
    const rejection = err => Promise.reject(err);
    const noop      = () => {};


    apiFoo.before(identity);
    apiFoo.after(identity, rejection);
    apiFoo.finally(noop);

    expect(apiFoo._middleware._before[0]).toBe(identity);
    expect(apiFoo._middleware._finally[0]).toBe(noop);
    expect(apiFoo._middleware._after[0]).toEqual({
      fulfill: identity,
      reject : rejection
    });
  });

  describe('before', () => {
    it('runs the before middlewares', () => {
      const apiFoo     = trae.create();
      const url        = `${TEST_URL}/foo`;
      let configRunned = false;

      fetchMock.mock(url, {
        status : 200,
        headers: {
          'Content-Type': 'application/json'
        },
        body: {
          foo: 'bar'
        }
      });

      apiFoo.before((config) => {
        config.headers.Authorization = '12345Foo';
        configRunned = true;
        return config;
      });

      return apiFoo.get(url)
      .then((res) => {
        expect(configRunned).toBe(true);
        expect(res).toMatchSnapshot();
        expect(fetchMock.called(url)).toBeTruthy();
        expect(fetchMock.lastUrl()).toBe(url);
        expect(fetchMock.lastOptions().method).toBe('GET');
        expect(fetchMock.lastOptions().headers.Authorization).toBe('12345Foo');
      });
    });
  });

  describe('after', () => {
    it('runs the fulfill after middlewares', () => {
      const apiFoo    = trae.create();
      const url       = `${TEST_URL}/foo`;
      let afterRunned = false;

      fetchMock.mock(url, {
        status : 200,
        headers: {
          'Content-Type': 'application/json'
        },
        body: {
          foo: 'bar'
        }
      });

      apiFoo.after((res) => {
        res.data.test = true;
        afterRunned   = true;
        return res;
      });

      return apiFoo.get(url)
      .then((res) => {
        expect(res.data.test).toBe(true);
        expect(afterRunned).toBe(true);
        expect(res).toMatchSnapshot();
        expect(fetchMock.called(url)).toBeTruthy();
        expect(fetchMock.lastUrl()).toBe(url);
        expect(fetchMock.lastOptions().method).toBe('GET');
      });
    });

    it('runs the reject after middlewares', () => {
      const apiFoo = trae.create();
      const url    = `${TEST_URL}/foo`;

      fetchMock.mock(url, {
        status: 500,
        body: { message: 'Error in the server' },
        headers: {
          'Content-Type': 'application/json'
        }
      });

      const fulfill = res => Promise.resolve(res);

      const reject = (err) => {
        err.test = true;
        return Promise.reject(err);
      };

      apiFoo.after(fulfill, reject);

      return apiFoo.get(url)
      .catch((err) => {
        expect(err.status).toBe(500);
        expect(err.test).toBe(true);
        expect(err.data).toEqual({ message: 'Error in the server' });
        expect(err).toMatchSnapshot();
        expect(fetchMock.called(url)).toBeTruthy();
        expect(fetchMock.lastUrl()).toBe(url);
        expect(fetchMock.lastOptions().method).toBe('GET');
      });
    });
  });

  describe('finally', () => {
    it('runs the finally middlewares', () => {
      const apiFoo      = trae.create();
      const url         = `${TEST_URL}/foo`;
      let finallyRunned = false;
      const mockConfig = {
        status : 200,
        headers: {
          'Content-Type': 'application/json'
        },
        body: {
          foo: 'bar'
        }
      };
      fetchMock.mock(url, mockConfig);

      apiFoo.finally((conf, passedUrl) => {
        expect(conf.mySpecialConfig).toBe('unicorn');
        expect(passedUrl).toBe(url);
        finallyRunned = true;
      });

      return apiFoo.get(url, { mySpecialConfig: 'unicorn' })
      .then((res) => {
        expect(finallyRunned).toBe(true);
        expect(res).toMatchSnapshot();
        expect(fetchMock.called(url)).toBeTruthy();
        expect(fetchMock.lastUrl()).toBe(url);
        expect(fetchMock.lastOptions().method).toBe('GET');
      });
    });
  });
});
