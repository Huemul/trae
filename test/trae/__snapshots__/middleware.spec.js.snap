// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`trae - middleware after runs the fulfill after middlewares 1`] = `
Object {
  "config": Object {
    "headers": Object {},
    "method": "GET",
    "url": "http://localhost:8080/api/foo",
  },
  "data": Object {
    "foo": "bar",
    "test": true,
  },
  "headers": Headers {
    Symbol(map): Object {
      "Content-Length": Array [
        "13",
      ],
      "Content-Type": Array [
        "application/json",
      ],
    },
  },
  "status": 200,
  "statusText": "OK",
}
`;

exports[`trae - middleware after runs the reject after middlewares 1`] = `[Error: Internal Server Error]`;

exports[`trae - middleware before runs the before middlewares 1`] = `
Object {
  "config": Object {
    "headers": Object {
      "Authorization": "12345Foo",
    },
    "method": "GET",
    "url": "http://localhost:8080/api/foo",
  },
  "data": Object {
    "foo": "bar",
  },
  "headers": Headers {
    Symbol(map): Object {
      "Content-Length": Array [
        "13",
      ],
      "Content-Type": Array [
        "application/json",
      ],
    },
  },
  "status": 200,
  "statusText": "OK",
}
`;

exports[`trae - middleware finally runs the finally middlewares 1`] = `
Object {
  "config": Object {
    "headers": Object {},
    "method": "GET",
    "mySpecialConfig": "unicorn",
    "url": "http://localhost:8080/api/foo",
  },
  "data": Object {
    "foo": "bar",
  },
  "headers": Headers {
    Symbol(map): Object {
      "Content-Length": Array [
        "13",
      ],
      "Content-Type": Array [
        "application/json",
      ],
    },
  },
  "status": 200,
  "statusText": "OK",
}
`;
